itests/kit2/ensemble.go

```diff
@@ -0,0 +1,602 @@
+		} else {
+			opts = append(opts, node.Override(new(modules.Genesis), modules.LoadGenesis(n.genesisBlock.Bytes())))
+		}
+
+		// Are we mocking proofs?
+		if n.options.mockProofs {
+			opts = append(opts,
+				node.Override(new(ffiwrapper.Verifier), mock.MockVerifier),
+				node.Override(new(ffiwrapper.Prover), mock.MockProver),
+			)
+		}
+
+		// Construct the full node.
+		stop, err := node.New(ctx, opts...)
+
+		// fullOpts[i].Opts(fulls),
+		require.NoError(n.t, err)
+
+		addr, err := full.WalletImport(context.Background(), &full.DefaultKey.KeyInfo)
+		require.NoError(n.t, err)
+
+		err = full.WalletSetDefault(context.Background(), addr)
+		require.NoError(n.t, err)
+
+		// Are we hitting this node through its RPC?
+		if full.options.rpc {
+			withRPC := fullRpc(n.t, full)
+			n.inactive.fullnodes[i] = withRPC
+		}
+
+		n.t.Cleanup(func() { _ = stop(context.Background()) })
+
+		n.active.fullnodes = append(n.active.fullnodes, full)
+	}
+
+	// If we are here, we have processed all inactive fullnodes and moved them
+	// to active, so clear the slice.
+	n.inactive.fullnodes = n.inactive.fullnodes[:0]
+
+	// Link all the nodes.
+	err := n.mn.LinkAll()
+	require.NoError(n.t, err)
+
+	// ---------------------
+	//  MINERS
+	// ---------------------
+
+	// Create all inactive miners.
+	for i, m := range n.inactive.miners {
+		if n.bootstrapped {
+			// this is a miner created after genesis, so it won't have a preseal.
+			// we need to create it on chain.
+			params, aerr := actors.SerializeParams(&power2.CreateMinerParams{
+				Owner:         m.OwnerKey.Address,
+				Worker:        m.OwnerKey.Address,
+				SealProofType: n.options.proofType,
+				Peer:          abi.PeerID(m.Libp2p.PeerID),
+			})
+			require.NoError(n.t, aerr)
+
+			createStorageMinerMsg := &types.Message{
+				From:  m.OwnerKey.Address,
+				To:    power.Address,
+				Value: big.Zero(),
+
+				Method: power.Methods.CreateMiner,
+				Params: params,
+
+				GasLimit:   0,
+				GasPremium: big.NewInt(5252),
+			}
+			signed, err := m.FullNode.FullNode.MpoolPushMessage(ctx, createStorageMinerMsg, nil)
+			require.NoError(n.t, err)
+
+			mw, err := m.FullNode.FullNode.StateWaitMsg(ctx, signed.Cid(), build.MessageConfidence, api.LookbackNoLimit, true)
+			require.NoError(n.t, err)
+			require.Equal(n.t, exitcode.Ok, mw.Receipt.ExitCode)
+
+			var retval power2.CreateMinerReturn
+			err = retval.UnmarshalCBOR(bytes.NewReader(mw.Receipt.Return))
+			require.NoError(n.t, err, "failed to create miner")
+
+			m.ActorAddr = retval.IDAddress
+		}
+
+		has, err := m.FullNode.WalletHas(ctx, m.OwnerKey.Address)
+		require.NoError(n.t, err)
+
+		// Only import the owner's full key into our companion full node, if we
+		// don't have it still.
+		if !has {
+			_, err = m.FullNode.WalletImport(ctx, &m.OwnerKey.KeyInfo)
+			require.NoError(n.t, err)
+		}
+
+		// // Set it as the default address.
+		// err = m.FullNode.WalletSetDefault(ctx, m.OwnerAddr.Address)
+		// require.NoError(n.t, err)
```

Comment:
should we remove this commented out code?

Reply:
I'm not entirely sure, I think we might need it for unmigrated tests? Let's take the plunge once we migrate them all!